tools/s3-gc-stats

```diff
@@ -0,0 +1,184 @@
+#!/usr/bin/env python3
+
+# Assumptions made by this code:
+# - list_object_versions sorts all versions (including delete markers) by key
+#   (in ascending order) and last modified date (in descending order, sometimes
+#   with consecutive equal timestamps) and returns them in chunks of 1000 (by
+#   default), with each chunk divided into proper versions and delete markers.
+# - Objects that have been deleted will have a delete marker as their latest
+#   version and at least one proper version dated before that.
+
+__requires__ = ["boto3", "click >= 7.0", "humanize"]
+
+from bisect import bisect
+from datetime import datetime
+import re
+import sys
+from typing import List, NamedTuple, Tuple
+from urllib.parse import urlparse
+
+import boto3
+from botocore import UNSIGNED
+from botocore.client import Config
+import click
+from humanize import naturalsize
+
+
+class Version(NamedTuple):
+    bucket: str
+    key: str
+    version_id: str
+    size: int
+    last_modified: datetime
+
+    @classmethod
+    def from_data(cls, bucket, data):
+        return cls(
+            bucket=bucket,
+            key=data["Key"],
+            version_id=data["VersionId"],
+            size=data["Size"],
+            last_modified=data["LastModified"],
+        )
+
+    @property
+    def key_url(self):
+        return f"s3://{self.bucket}/{self.key}"
+
+    @property
+    def url(self):
+        return f"s3://{self.bucket}/{self.key}?versionId={self.version_id}"
+
+    def __str__(self):
+        return f"{self.url} {self.size}"
+
+
+class BucketStats:
+    def __init__(self, bucket, prefix, list_files=False, stat=("all",), exclude=()):
+        self.bucket: str = bucket
+        self.prefix: str = prefix
+        self.list_files: bool = list_files
+        self.stat: Tuple[str, ...] = stat
+        self.exclude: Tuple[str, ...] = exclude
+        #: Versions of the current key
+        self.versions: List[Version] = []
+        # Deleted keys, in ascending order:
+        self.deleted: List[str] = []
+        self.qtys = {
+            "all": 0,
+            "visible": 0,
+            "invisible": 0,
+            "old": 0,
+        }
+        self.sizes = {
+            "all": 0,
+            "visible": 0,
+            "invisible": 0,
+            "old": 0,
+        }
+        self.found_any = False
+
+    def run(self):
+        # Use s3 anonymously/without credentials:
+        client = boto3.client("s3", config=Config(signature_version=UNSIGNED))
+        for page in client.get_paginator("list_object_versions").paginate(
+            Bucket=self.bucket, Prefix=self.prefix
+        ):
+            # TODO: Filter out keys that ends with slashes?
+            for dm in page.get("DeleteMarkers", []):
+                if dm["IsLatest"]:
+                    self.mark_deleted(dm["Key"])
+            for v in page["Versions"]:
```

Comment:
<details>
<summary>note: it seems to correctly work out the keys which were added before versioning was turned on
 -- look at 1version-nonversioned1.txt and 2versions-nonversioned1.txt . GREAT!</summary> 

```shell
(git)lena:~/proj/dandi/dandi-cli-master[gh-470]git
$> tools/s3-gc-stats --stat visible --list s3://datalad-test0-versioned
s3://datalad-test0-versioned/1version-nonversioned1.txt?versionId=null 8
s3://datalad-test0-versioned/1version-removed-recreated.txt?versionId=vurmCZo5d3.iAzTV.YAfTcMr.LDVjQIK 8
s3://datalad-test0-versioned/2versions-nonversioned1.txt?versionId=V4Dqhu0QTEtxmvoNkCHGrjVZVomR1Ryo 8
s3://datalad-test0-versioned/2versions-nonversioned1.txt_sameprefix?versionId=QUOLbSrBHBmU3UNeEvdfcX2puqRffH9l 8
s3://datalad-test0-versioned/2versions-removed-recreated.txt?versionId=bBVSCB4MdBOeEXDQ2KwrjtrevpwFabaY 8
s3://datalad-test0-versioned/2versions-removed-recreated.txt_sameprefix?versionId=cfTdf3N8exZLFg.KcW5szQKrFNLUyCu1 8
s3://datalad-test0-versioned/3versions-allversioned.txt?versionId=Kvuind11HZh._dCPaDAb0OY9dRrQoTMn 8
s3://datalad-test0-versioned/3versions-allversioned.txt_sameprefix?versionId=Mvsc4FgJWc6gExwSw1d6wsLrnk6wdDVa 8
Visible files: 8 / Size: 64 (64 Bytes)
1 66166.....................................:Wed 17 Mar 2021 10:34:34 AM EDT:.
(git)lena:~/proj/dandi/dandi-cli-master[gh-470]git
$> tools/s3-gc-stats --stat invisible --list s3://datalad-test0-versioned
s3://datalad-test0-versioned/1version-removed-recreated.txt?versionId=B7cP2PKGRNeRZ7EktWN82UnUaGjb9dh_ 8
s3://datalad-test0-versioned/1version-removed.txt?versionId=eZ5Hgwo8azfBv3QT7aW9dmm2sbLUY.QP 8
s3://datalad-test0-versioned/2versions-nonversioned1.txt?versionId=null 8
s3://datalad-test0-versioned/2versions-removed-recreated.txt?versionId=zwW0b567gYO3puJeLMZsOmETqowJnv6l 8
s3://datalad-test0-versioned/3versions-allversioned.txt?versionId=b.qCuh7Sg58VIYj8TVHzbRS97EvejzEl 8
s3://datalad-test0-versioned/3versions-allversioned.txt?versionId=pNsV5jJrnGATkmNrP8.i_xNH6CY4Mo5s 8
Invisible files: 6 / Size: 48 (48 Bytes)

```
</details>